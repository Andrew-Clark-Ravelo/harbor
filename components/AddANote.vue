<script setup>
import { computed, ref } from "vue";
import { useSingleReport } from "../composables";

const props = defineProps({
  name: {
    type: String,
    required: true,
  },
});

const tempFormValue = ref({});
const { addNote, report, isLoadingNote } = useSingleReport();
const currentPackage = computed(() => report.value.state?.[props.name] || {});

const updateFormText = (evt) => {
  tempFormValue.value[props.name] = evt.target.value;
};

const noteInnerText = computed(() => {
  return tempFormValue.value[props.name] || currentPackage.value.note;
});

const noteText = computed(() =>
  currentPackage.value.note
    ? "update note for this library"
    : "add a note for this library"
);

const notePlaceholder = "add note";
const buttonText = computed(() => (isLoadingNote.value ? "..." : "+"));
</script>

<template>
  <details class="mt-8 cursor-pointer text-axo-orange text-xl">
    <summary class="">{{ noteText }}</summary>
    <div
      class="mt-8 w-4/5 bg-slate-800 border-t border-t-slate-600 flex justify-between gap-4 text-slate-50 px-4 pb-6 pt-2 text-xs"
    >
      <div class="w-full">
        <label for="comment" class="sr-only">{{ noteText }}</label>
        <div class="flex gap-4 items-end">
          <textarea
            id="comment"
            :value="noteInnerText"
            :placeholder="notePlaceholder"
            rows="1"
            name="comment"
            class="bg-transparent outline-none border-0 w-full p-4 border-b-axo-orange border-b active:border-b-axo-pink focus:border-b-axo-pink focus:border-b-2 focus:outline-none focus:ring-0 placeholder:text-slate-400"
            @input="updateFormText"
          />
          <button
            class="min-w-max bg-axo-orange text-xl hover:bg-axo-pink px-4 py-2 rounded-md"
            @click="addNote({ pkg: props.name, note: tempFormValue })"
          >
            {{ buttonText }}
          </button>
        </div>
      </div>
    </div>
  </details>
</template>
