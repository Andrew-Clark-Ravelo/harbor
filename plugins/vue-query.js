import { VueQueryPlugin, QueryClient, hydrate } from "@tanstack/vue-query";

export default defineNuxtPlugin((nuxtApp) => {
  // Modify your Vue Query global settings here
  const queryClient = new QueryClient({
    defaultOptions: { queries: { staleTime: Infinity } },
  });
  const options = { queryClient };

  nuxtApp.vueApp.use(VueQueryPlugin, options);

  if (process.client) {
    if (nuxtApp.payload.state && nuxtApp.payload.state["vueQueryState"]) {
      hydrate(queryClient, nuxtApp.payload.state["vueQueryState"]);
    }
  }
});
