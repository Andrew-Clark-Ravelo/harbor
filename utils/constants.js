export const GH_TOKEN = "gh_token";
export const GH_STATE = "gh_user";
export const REDIRECT_OATH_URL = `https://github.com/login/oauth/authorize?scope=repo&client_id=`;
export const REDIRECT_COOKIE = "redirect_to";

export const MISSING_CRITERIA_KEYS = {
  "safe-to-deploy": {
    name: "Safe to deploy",
    "description-url":
      "https://raw.githubusercontent.com/mozilla/cargo-vet/main/src/criteria/safe-to-deploy.txt",
  },
  "safe-to-run": {
    name: "Safe to run",
    "description-url":
      "https://raw.githubusercontent.com/mozilla/cargo-vet/main/src/criteria/safe-to-run.txt",
  },
};
