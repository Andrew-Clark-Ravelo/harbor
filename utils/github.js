export const postGh = ({ url, cookie, body }) => {
  return $fetch(url, {
    method: "POST",
    baseURL: "https://api.github.com",
    body,
    headers: {
      accept: "application/vnd.github+json",
      Authorization: `Bearer ${cookie}`,
    },
  });
};

export const fetchGh = ({ url, cookie }) => {
  return $fetch(url, {
    baseURL: "https://api.github.com",
    headers: {
      Authorization: `token ${cookie}`,
    },
  });
};
