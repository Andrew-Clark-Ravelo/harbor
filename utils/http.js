import axios from "axios";

export const http = {
  get: async (url) => {
    const { data } = await axios.get(url, {
      headers: useRequestHeaders(["cookie"]),
    });
    return data;
  },
  post: async (url, body) => {
    const { data } = await axios.post(url, body);
    return data;
  },

  put: async (url, body) => {
    const { data } = await axios.put(url, body);
    return data;
  },
};
