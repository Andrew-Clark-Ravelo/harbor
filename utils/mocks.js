export const SINGLE_REPORT_MOCK = {
  suggestions: [
    {
      name: "clap",
      notable_parents: "cargo-vet-school",
      suggested_criteria: ["safe-to-deploy"],
      suggested_diff: {
        from: "4.0.15",
        to: "4.0.29",
        diffstat: {
          raw: " 28 files changed, 384 insertions(+), 175 deletions(-)\n",
          count: 559,
        },
      },
      confident: true,
    },
    {
      name: "serde_json",
      notable_parents: "reqwest, cargo-vet-school",
      suggested_criteria: ["safe-to-deploy"],
      suggested_diff: {
        from: null,
        to: "0.2.8",
        diffstat: {
          raw: " 15 files changed, 804 insertions(+)\n",
          count: 804,
        },
      },
      confident: true,
    },
  ],
  criteria: {
    "legalese-audit": {
      description: "a cool description for legalese-audit",
      implies: ["safe-to-run", "too-big-to-fail"],
    },
    "safe-to-deploy": {
      description: "a cool description for safe to deploy",
    },
  },
  state: {
    clap: {
      "safe-to-deploy": true,
      note: "one",
    },
    serde_json: {
      "safe-to-deploy": false,
    },
  },
};
