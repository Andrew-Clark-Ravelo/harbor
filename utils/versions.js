export const getVersionChangeText = (diff) => {
  return diff?.from
    ? `changed from v${diff?.from} to v${diff?.to}`
    : `full audit recommended`;
};
