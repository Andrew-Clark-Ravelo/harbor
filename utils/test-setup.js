import { setupServer } from "msw/node";
import { rest } from "msw";
import { SINGLE_REPORT_MOCK } from "./mocks";

const server = setupServer(
  rest.get("///api/reports/678-s", (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(SINGLE_REPORT_MOCK));
  }),
  rest.put("///api/reports/678-s", (req, res, ctx) => {
    return res(ctx.status(200), ctx.json({}));
  })
);

beforeAll(() => {
  const useRoute = vi.fn(() => ({
    params: { id: "678-s" },
    query: { name: "clap" },
  }));

  vi.stubGlobal("useRoute", useRoute);

  const useRequestHeaders = vi.fn(() => {});

  vi.stubGlobal("useRequestHeaders", useRequestHeaders);
  server.listen({ onUnhandledRequest: "error" });
});

afterAll(() => server.close());

afterEach(() => server.resetHandlers());
