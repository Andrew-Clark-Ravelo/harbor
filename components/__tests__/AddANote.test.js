import { mount, flushPromises } from "@vue/test-utils";
import AddANote from "../AddANote.vue";
import { expect } from "vitest";
import axios from "axios";
import { SINGLE_REPORT_MOCK } from "../../utils/mocks";
import { NOTE_MESSAGES } from "../../utils/constants";
import { VueQueryPlugin } from "@tanstack/vue-query";

let wrapper;

const createComponent = ({ props = {}, shallow = false } = {}) => {
  wrapper = mount(AddANote, {
    props,
    shallow,
    global: {
      plugins: [VueQueryPlugin],
    },
  });
};

const currentPkg = SINGLE_REPORT_MOCK.suggestions[1];
const getSummary = () => wrapper.find("summary");
const getTextarea = () => wrapper.find("textarea");

describe("AddANote component", () => {
  const axiosPut = vi.spyOn(axios, "put");
  beforeEach(async () => {
    createComponent({
      props: {
        name: currentPkg.name,
      },
    });
    await flushPromises();
  });

  afterEach(() => {
    vi.restoreAllMocks();
    wrapper.unmount();
    wrapper = null;
  });

  it("mounts and shows empty textarea", async () => {
    expect(getSummary().text()).toContain(NOTE_MESSAGES.empty);

    await getSummary().trigger("click");

    expect(getTextarea().value).toBeUndefined();

    await getTextarea().setValue("a comment");
    await wrapper.find("button").trigger("click");
    expect(axiosPut).toHaveBeenCalledWith("/api/reports/678", {
      ...SINGLE_REPORT_MOCK.state,
      [currentPkg.name]: {
        ...SINGLE_REPORT_MOCK.state[currentPkg.name],
        note: "a comment",
      },
    });
  });

  it("uses default value", async () => {
    await wrapper.setProps({
      name: SINGLE_REPORT_MOCK.suggestions[0].name,
    });
    expect(getSummary().text()).toBe(NOTE_MESSAGES.filled);
  });

  it("triggers re render on pkg change", async () => {
    expect(getSummary().text()).toBe(NOTE_MESSAGES.empty);

    await wrapper.setProps({
      name: SINGLE_REPORT_MOCK.suggestions[0].name,
    });
    expect(getSummary().text()).toBe(NOTE_MESSAGES.filled);
  });
});
