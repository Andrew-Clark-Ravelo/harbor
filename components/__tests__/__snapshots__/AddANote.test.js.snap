// Vitest Snapshot v1

exports[`mounts component 1`] = `
<div>
  <details
    class="mt-8 cursor-pointer text-axo-orange text-xl"
  >
    <summary
      class=""
    >
      add a note for this library
    </summary>
    <div
      class="mt-8 w-4/5 bg-slate-800 border-t border-t-slate-600 flex justify-between gap-4 text-slate-50 px-4 pb-6 pt-2 text-xs"
    >
      <div
        class="w-full"
      >
        <label
          class="sr-only"
          for="comment"
        >
          add a note for this library
        </label>
        <div
          class="flex gap-4 items-end"
        >
          <textarea
            class="bg-transparent outline-none border-0 w-full p-4 border-b-axo-orange border-b active:border-b-axo-pink focus:border-b-axo-pink focus:border-b-2 focus:outline-none focus:ring-0 placeholder:text-slate-400"
            id="comment"
            name="comment"
            placeholder="add note"
            rows="1"
          />
          <button
            class="min-w-max bg-axo-orange text-xl hover:bg-axo-pink px-4 py-2 rounded-md"
          >
            +
          </button>
        </div>
      </div>
    </div>
  </details>
</div>
`;

exports[`mounts component with default note 1`] = `
<div>
  <details
    class="mt-8 cursor-pointer text-axo-orange text-xl"
  >
    <summary
      class=""
    >
      update note for this library
    </summary>
    <div
      class="mt-8 w-4/5 bg-slate-800 border-t border-t-slate-600 flex justify-between gap-4 text-slate-50 px-4 pb-6 pt-2 text-xs"
    >
      <div
        class="w-full"
      >
        <label
          class="sr-only"
          for="comment"
        >
          update note for this library
        </label>
        <div
          class="flex gap-4 items-end"
        >
          <textarea
            class="bg-transparent outline-none border-0 w-full p-4 border-b-axo-orange border-b active:border-b-axo-pink focus:border-b-axo-pink focus:border-b-2 focus:outline-none focus:ring-0 placeholder:text-slate-400"
            id="comment"
            name="comment"
            placeholder="add note"
            rows="1"
          />
          <button
            class="min-w-max bg-axo-orange text-xl hover:bg-axo-pink px-4 py-2 rounded-md"
          >
            +
          </button>
        </div>
      </div>
    </div>
  </details>
</div>
`;
